File: tools/asm2nl.js
const ASM_FOLDER_PATH = "C:\\Lupin Sansei - Pandora no Isan\\";
const NL_FOLDER_PATH  = "C:\\Lupin Sansei - Pandora no Isan\\";
const ASM_FILE1 = "bank_FF.asm";
const ASM_FILE2 = "constants.inc";
const NL_FILE1 = "game.nes.7.nl";
const NL_FILE2 = "game.nes.ram.nl";
const SKIP_COMMERCIAL_AT = false;
const LOOP_WITH_COUNTER = false;
const COUNTER_LIMIT = 65;
const ONLY_LABEL = false;
const EXCLUDE_LIST = [ '0005', '0011', '0012', '001A', '0039', '003A' ];

function getBankNl()
{
    var array = [];
    var asmFile = FileSystem.openTextFile(ASM_FOLDER_PATH + ASM_FILE1, "utf8");
    try
    {
        var label = null;
        var comment = null;
        var counter = 0;
        while ((!LOOP_WITH_COUNTER || counter < COUNTER_LIMIT) && !asmFile.eof)
        {
            var line = asmFile.readLine();
            if (typeof line === 'string')
            {
                if (line.match("^.\\s.\\s.\\s.\\s.\\s.\\s0x0........:....:\\s..\\s..\\s..\\s\\s"))
                {
                    if (!ONLY_LABEL)
                    {
                        var parts = line.split(';');
                        if (parts.length > 1)
                        {
                            comment = Strings.trimLeft(parts[1]);
                            if (comment.length == 0)
                                comment = ';';
                        }
                    }

                    line = line.replace(/^.\s.\s.\s.\s.\s.\s0x0........:/, "");
                    var address = line[0] + line[1] + line[2] + line[3];
                    var value = parseInt(address, 16);

                    if (label != null || comment != null)
                    {
                        if (SKIP_COMMERCIAL_AT && label != null && label[0] == '@')
                            label = label.replace(/^\@/, "");

                        array[value] = {label: label, comment: comment};
                        label = null;
                        comment = null;
                    }
                }
                else
                {
                    var parts = line.split(':');
                    if (parts.length > 1)
                        label = parts[0];
                    else
                        label = null;
                }
            }
            counter++;
        }
    }
    finally
    {
        asmFile.close();
    }
    if (array.length == 0)
        return;

    var nlFile = FileSystem.createTextFile(NL_FOLDER_PATH + NL_FILE1, "utf8");
    try
    {
        for (var i = 0; i < array.length; i++)
        {
            if (typeof array[i] === 'undefined')
                continue;
            var value = Strings.format("%x", i);
            var label = '';
            var comment = '';
            if (array[i]['label'])
                label = array[i]['label'];
            if (array[i]['comment'])
                comment = array[i]['comment'];
            nlFile.writeLine("$"+value+"#"+label+"#"+comment);
        }
        //nlFile.writeLine();
    }
    finally
    {
        nlFile.close();
    }
}

function getVariablesNl()
{
    var array = [];
    var asmFile = FileSystem.openTextFile(ASM_FOLDER_PATH + ASM_FILE2, "utf8");
    try
    {
        var counter = 0;
        while ((!LOOP_WITH_COUNTER || counter < COUNTER_LIMIT) && !asmFile.eof)
        {
            var line = asmFile.readLine();
            if (typeof line === 'string')
            {
                var ramValue = null;
                var parts = line.split('=');
                if (parts.length > 1)
                    ramValue = Strings.trimLeft(parts[1]);
                if (ramValue != null && ramValue.match("^ram_"))
                {
                    ramValue = ramValue.replace(/^ram_/, "");
                    var address = ramValue[0] + ramValue[1] + ramValue[2] + ramValue[3];
                    var value = parseInt(address, 16);

                    var label = Strings.trimAll(parts[0]);
                    if (label && label.length > 0)
                        array[value] = {label: label};
                }
            }
            counter++;
        }
    }
    finally
    {
        asmFile.close();
    }
    if (array.length == 0)
        return;

    var nlFile = FileSystem.createTextFile(NL_FOLDER_PATH + NL_FILE2, "utf8");
    try
    {
        for (var i = 0; i < array.length; i++)
        {
            if (typeof array[i] === 'undefined')
                continue;
            var value = Strings.format("%x", i);
            if (value.length == 1)
                value = "000" + value;
            if (value.length == 2)
                value = "00" + value;
            if (value.length == 3)
                value = "0" + value;

            var found = false;
            for (var j = 0; j < array.length; j++)
            {
                if (EXCLUDE_LIST[j] == value)
                {
                    found = true;
                    break;
                }
            }
            if (found)
                continue;

            var label = array[i]['label'];
            nlFile.writeLine("$"+value+"#"+label+"#");
        }
        //nlFile.writeLine();
    }
    finally
    {
        nlFile.close();
    }
}

File: tools/asm2nl.py
"""
Python 3.7.4
"""

import re

ASM_FOLDER_PATH = "C:\\Lupin Sansei - Pandora no Isan\\"
NL_FOLDER_PATH  = "C:\\Lupin Sansei - Pandora no Isan\\"
ASM_FILE1 = "bank_FF.asm"
ASM_FILE2 = "constants.inc"
NL_FILE1 = "game.nes.7.nl"
NL_FILE2 = "game.nes.ram.nl"
SKIP_COMMERCIAL_AT = True
LOOP_WITH_COUNTER = False
COUNTER_LIMIT = 65
ONLY_LABEL = False
EXCLUDE_LIST = [ '0005', '0011', '0012', '001A', '0039', '003A' ]

def getBankNl(asmNameFile, nlNameFile):
    dictionary = {}
    asmFile = open(ASM_FOLDER_PATH + asmNameFile, encoding="utf-8")
    try:
        label = None
        comment = None
        counter = 0
        for line in asmFile:
            if LOOP_WITH_COUNTER and counter >= COUNTER_LIMIT:
                break
            if re.search("^.\\s.\\s.\\s.\\s.\\s.\\s0x0........:....:\\s..\\s..\\s..\\s\\s", line):
                if not ONLY_LABEL:
                    parts = line.split(';')
                    if len(parts) > 1:
                        comment = parts[1].strip()
                        if len(comment) == 0:
                            comment = '--NO-COMMENT--'
                line = re.sub("^.\s.\s.\s.\s.\s.\s0x0........:", "", line)
                address = line[0] + line[1] + line[2] + line[3]
                value = int(address, base=16)

                if (label is not None) or (comment is not None):
                    if SKIP_COMMERCIAL_AT and (label is not None) and label[0] == '@':
                        label = re.sub("^\@", "", label)

                    dictionary[str(value)] = {"label": label, "comment": comment}
                    label = None
                    comment = None
            else:
                parts = line.split(':')
                if len(parts) > 1:
                    label = parts[0]
                else:
                    label = None
            counter += 1
    finally:
        asmFile.close()
    if len(dictionary) == 0:
        return None

    nlFile = open(NL_FOLDER_PATH + nlNameFile, "w", encoding="utf-8")
    try:
        counter = len(dictionary)
        for key, value in dictionary.items():
            counter -= 1
            iValue = "{0:X}".format(int(key))
            label = ''
            comment = ''
            if value['label']:
                label = value['label']
            if value['comment']:
                comment = value['comment']
            if counter == 0:
                nlFile.write("$"+iValue+"#"+label+"#"+comment)
            else:
                nlFile.write("$"+iValue+"#"+label+"#"+comment+"\n")
    finally:
        nlFile.close()

def getVariablesNl():
    dictionary = {}
    asmFile = open(ASM_FOLDER_PATH + ASM_FILE2, encoding="utf-8")
    try:
        counter = 0
        for line in asmFile:
            if LOOP_WITH_COUNTER and counter >= COUNTER_LIMIT:
                break
            ramValue = None
            parts = line.split('=')
            if len(parts) > 1:
                ramValue = parts[1].strip()

            if (ramValue is not None) and re.search("^ram_", ramValue):
                ramValue = re.sub("^ram_", "", ramValue)
                address = ramValue[0] + ramValue[1] + ramValue[2] + ramValue[3]
                value = int(address, base=16)
                label = parts[0].strip()
                if len(label) > 0:
                    dictionary[str(value)] = {"label": label}
            counter += 1
    finally:
        asmFile.close()
    if len(dictionary) == 0:
        return None

    nlFile = open(NL_FOLDER_PATH + NL_FILE2, "w", encoding="utf-8")
    try:
        counter = len(dictionary)
        for key, value in dictionary.items():
            counter -= 1
            iValue = "{0:X}".format(int(key))

            if len(iValue) == 1:
                iValue = "000" + iValue
            elif len(iValue) == 2:
                iValue = "00" + iValue
            elif len(iValue) == 3:
                iValue = "0" + iValue
            if iValue in EXCLUDE_LIST:
                continue

            label = value['label']
            if counter == 0:
                nlFile.write("$"+iValue+"#"+label+"#")
            else:
                nlFile.write("$"+iValue+"#"+label+"#\n")
    finally:
        nlFile.close()

def getBankFFNl():
    getBankNl(ASM_FILE1, NL_FILE1)

getBankFFNl()
getVariablesNl()
